src/azul/drs.py

```diff
@@ -74,10 +87,158 @@ class Access:
     headers: Optional[Mapping[str, str]] = None
 
 
+class DRSURI(metaclass=ABCMeta):
+
+    @property
+    @abstractmethod
+    def _object_id(self) -> str:
+        """
+        The portion of the URI that identifies the data object.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def parse(cls, drs_uri: str) -> 'DRSURI':
+        try:
+            drs_uri = RegularDRSURI.parse(drs_uri)
+        except ValueError as e:
+            if len(e.args) == 1 and isinstance(e.args[0], str) and e.args[0].startswith('Invalid port '):
+                drs_uri = CompactDRSURI.parse(drs_uri)
+            else:
+                raise
+        return drs_uri
+
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        """
+        Translate the DRS URI into a DRS URL. All query params included in the DRS
+        URI (eg '{drs_uri}?version=123') will be carried over to the DRS URL.
+        """
+        url = self._url_origin(client)
+        url.set(path=drs_object_url_path(object_id=self._object_id, access_id=access_id))
+        # "The colon character is not allowed in a hostname-based DRS URI".
+        # https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_uris
+        reject(':' in url.netloc or ':' in str(url.path),
+               f'The DRS URL {url!r} is not hostname-based')
+        return str(url)
+
+    @abstractmethod
+    def _url_origin(self, client: 'DRSClient') -> mutable_furl:
```

Comment:
As I said, abstract methods, protected or public, need a docstring, simply because there is no body to learn anything from.